prov/efa/src/rdm/efa_rdm_ope.c

```diff
@@ -20,56 +20,47 @@ void efa_rdm_txe_construct(struct efa_rdm_ope *txe,
 {
 	uint64_t tx_op_flags;
 
-	txe->ep = ep;
-	txe->type = EFA_RDM_TXE;
-	txe->op = op;
-	txe->tx_id = ofi_buf_index(txe);
-	txe->state = EFA_RDM_TXE_REQ;
-	txe->addr = msg->addr;
-	txe->peer = peer;
+	/* set flags */
+	assert(ep->base_ep.util_ep.tx_msg_flags == 0 ||
+	       ep->base_ep.util_ep.tx_msg_flags == FI_COMPLETION);
+	tx_op_flags = ep->base_ep.util_ep.tx_op_flags;
+	if (ep->base_ep.util_ep.tx_msg_flags == 0)
+		tx_op_flags &= ~FI_COMPLETION;
+
+	*txe = (struct efa_rdm_ope) {
```

Comment:
yikes, this is a huge structure that we're now initializing to zero.  I'm not really sure I understand why the structure is so large, but it's concerning that every operation starts with initializing something that big.  This is not reasonable for a zero-copy path in particular.

Reply:
Zero-initializing was the point of the refactor here. I find the alternative prone to bugs and less maintainable.

To be clear: I'm guessing the storage duration/stack utilization is the primary concern here, correct? I hadn't really considered that - likely explains the existing `memset()` usage.

Reply:
memset(0) on a big region is also a problem.  The big problem is you have a large structure that you're carrying around per operation.  In the zero copy case, you shouldn't need anything more than some way of getting back to the context to return to the user.  You've basically made it so that the fast path is limited by the slow path.

Readability / maintainability is important, but performance is the real goal.